### PR TITLE
Replace direct usage of p:d:t by DistributedTriangulationBase

### PR DESCRIPTION
**Pull request by peterrum**
As a preparation to introduce `parallel::fullydistributed::Triangulation`.

> 100% tests passed, 0 tests failed out of 30

### CODE REVIEW

File: include/solvers/navier_stokes_base.h [hunk 1]
```diff
@@ -376,11 +376,13 @@ NavierStokesBase<dim, VectorType, DofsType>::NavierStokesBase(
   , mpi_communicator(MPI_COMM_WORLD)
   , n_mpi_processes(Utilities::MPI::n_mpi_processes(mpi_communicator))
   , this_mpi_process(Utilities::MPI::this_mpi_process(mpi_communicator))
-  , triangulation(this->mpi_communicator,
-                  typename Triangulation<dim>::MeshSmoothing(
-                    Triangulation<dim>::smoothing_on_refinement |
-                    Triangulation<dim>::smoothing_on_coarsening))
-  , dof_handler(this->triangulation)
+  , triangulation(dynamic_cast<parallel::DistributedTriangulationBase<dim> *>(
```
Comment (peterrum) [inline include/solvers/navier_stokes_base.h]: This is the place where an instance of `p:f:t` will be created.

File: include/solvers/navier_stokes_base.h [hunk 2]
```diff
@@ -1266,11 +1268,18 @@ template <int dim, typename VectorType, typename DofsType>
 void
 NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
 {
+  if (dynamic_cast<parallel::distributed::Triangulation<dim> *>(
```
Comment (peterrum) [inline include/solvers/navier_stokes_base.h]: Adaptive mesh refinement is unfortunately not supported at the moment for `p:f:t`. However, there is already some work in that direction (see https://github.com/dealii/dealii/pull/8834).

File: include/solvers/navier_stokes_base.h [hunk 3]
```diff
@@ -1586,7 +1592,9 @@ NavierStokesBase<dim, VectorType, DofsType>::read_checkpoint()
 
   try
     {
-      this->triangulation.load(filename.c_str());
+      if (auto tria = dynamic_cast<parallel::distributed::Triangulation<dim> *>(
+            this->triangulation.get()))
+        tria->load(filename.c_str());
```
Comment (peterrum) [inline include/solvers/navier_stokes_base.h]: The function `load` is also not implemented yet...

File: include/solvers/navier_stokes_base.h [hunk 4]
```diff
@@ -1883,8 +1891,13 @@ NavierStokesBase<dim, VectorType, DofsType>::write_checkpoint()
...
   system_trans_vectors.prepare_for_serialization(sol_set_transfer);
 
-  std::string triangulationName = prefix + ".triangulation";
-  this->triangulation.save(prefix + ".triangulation");
+
+  if (auto tria = dynamic_cast<parallel::distributed::Triangulation<dim> *>(
+        this->triangulation.get()))
+    {
+      std::string triangulationName = prefix + ".triangulation";
+      tria->save(prefix + ".triangulation");
```
Comment (peterrum) [inline include/solvers/navier_stokes_base.h]: The same for `save`...

File: include/solvers/navier_stokes_base.h (hunk 1 above)
Comment (blaisb) [reply include/solvers/navier_stokes_base.h]: Seems coherent

**Comment by blaisb** [comment]
So if I get this one correctly, right now you only made it so that the triangulation is a shared vector to the base parallel distributed triangulation class (which is aimed at supporting both types of parallel triangulations)?
All test pass so i'll wait for your OK and then merge to master.
The code is going to undergo some change in the next few weeks to make it more readable and just improve the design.
Waiting for your go and then I'll merge !

**Comment by peterrum** [comment]
Exactly! Ready to merge!

**Comment by blaisb** [comment]
Done
Thanks! This will be a great addition.
We'll be coming up with a paper soon, we should keep in touch for that.

**Pull request merged**